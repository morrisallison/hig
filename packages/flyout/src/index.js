import Flyout from "./Flyout";
import Panel from "./presenters/PanelContainerPresenter";

Flyout.Panel = Panel;

export { Flyout as default, Panel };
export {
  anchorPoints,
  AVAILABLE_ANCHOR_POINTS,
  /** @todo Remove alias */
  AVAILABLE_ANCHOR_POINTS as availableAnchorPoints
} from "./anchorPoints";
export {
  offsetContainerVertical,
  offsetPanelHorizontal
} from "./coordinateHelpers";
