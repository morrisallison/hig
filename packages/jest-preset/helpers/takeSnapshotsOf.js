const React = require("react");
const renderer = require("react-test-renderer");

/**
 * @typedef {import("react-test-renderer").TestRendererOptions} Options
 */

/**
 * @typedef {Object} Case
 * @property {string} desc
 * @property {Object} [props]
 * @property {import("react").ReactElement} [subject]
 */

/** @type {Options} */
const defaultOptions = {
  createNodeMock(element) {
    return document.createElement(element.type);
  }
};

/**
 * @param {import("react").Component|null} Component
 * @param {Case[]} cases
 * @param {Options} [options]
 */
module.exports = function takeSnapshotsOf(Component, cases, options) {
  cases.forEach(({ desc, props, subject }) => {
    it(desc, () => {
      const rendererOptions = Object.assign({}, defaultOptions, options);
      const vDom = subject || React.createElement(Component, props);
      const tree = renderer.create(vDom, rendererOptions).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
};
