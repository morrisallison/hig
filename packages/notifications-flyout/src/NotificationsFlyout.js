import React from "react";
import PropTypes from "prop-types";

import Flyout, { AVAILABLE_ANCHOR_POINTS } from "@hig/flyout";
import "@hig/flyout/build/index.css";

import NotificationFlyoutBehavior from "./behaviors/NotificationFlyoutBehavior";
import IndicatorPresenter from "./presenters/IndicatorPresenter";

import Notification from "./Notification";
import Panel from "./Panel";

function createNotificationRenderer({ hideFlyout, dismissNotification }) {
  /* eslint-disable-next-line react/prop-types */
  return function renderNotification({ id, key, content, ...otherProps }) {
    return (
      <Notification
        {...otherProps}
        hideFlyout={hideFlyout}
        id={id}
        key={key}
        onDismiss={() => dismissNotification(id)}
      >
        {content}
      </Notification>
    );
  };
}

function createPanelRenderer({
  dismissNotification,
  notifications,
  loading,
  heading
}) {
  /* eslint-disable-next-line react/prop-types */
  return function renderPanel({ hideFlyout, handleScroll, innerRef }) {
    return (
      <Panel
        heading={heading}
        innerRef={innerRef}
        loading={loading}
        onScroll={handleScroll}
      >
        {notifications.map(
          createNotificationRenderer({ hideFlyout, dismissNotification })
        )}
      </Panel>
    );
  };
}

export default function NotificationsFlyout(props) {
  const {
    anchorPoint,
    heading,
    indicatorTitle,
    loading,
    onClickOutside,
    onScroll,
    open,
    children,
    unreadCount: controlledUnreadCount,
    notifications: rawNotifications = children
  } = props;

  return (
    <NotificationFlyoutBehavior
      unreadCount={controlledUnreadCount}
      notifications={rawNotifications}
    >
      {({
        dismissNotification,
        handleClose,
        notifications,
        showUnreadCount,
        unreadCount
      }) => (
        <Flyout
          anchorPoint={anchorPoint}
          onClickOutside={onClickOutside}
          onClose={handleClose}
          onScroll={onScroll}
          open={open}
          panel={createPanelRenderer({
            dismissNotification,
            notifications,
            loading,
            heading
          })}
        >
          {({ handleClick }) => (
            <IndicatorPresenter
              onClick={handleClick}
              count={unreadCount}
              showCount={showUnreadCount}
              title={indicatorTitle}
            />
          )}
        </Flyout>
      )}
    </NotificationFlyoutBehavior>
  );
}

NotificationsFlyout.propTypes = {
  /** Where the flyout will be anchored relative to target */
  anchorPoint: PropTypes.oneOf(AVAILABLE_ANCHOR_POINTS),
  /** Rendered notifications */
  children: PropTypes.node,
  /** Flyout panel heading */
  heading: PropTypes.string,
  /** Indicator button title */
  indicatorTitle: PropTypes.string,
  /** Determines whether the loading animation is shown */
  loading: PropTypes.bool,
  /** Rendered notifications */
  notifications: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.node, PropTypes.shape({})])
  ),
  /** Function called when the flyout is open, and a click event occurs outside the flyout */
  onClickOutside: PropTypes.func,
  /** Function called when the flyout panel is scrolled */
  onScroll: PropTypes.func,
  /** When provided, it overrides the flyout's open state */
  open: PropTypes.bool,
  /** When provided, it overrides the derived unread notification count */
  unreadCount: PropTypes.number
};
