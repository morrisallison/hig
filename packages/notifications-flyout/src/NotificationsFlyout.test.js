import { anchorPoints } from "@hig/flyout";
import { takeSnapshotsOf } from "@hig/jest-preset/helpers";

import NotificationsFlyout from "./NotificationsFlyout";

describe("notifications-flyout/NotificationsFlyout", () => {
  takeSnapshotsOf(NotificationsFlyout, [
    {
      desc: "renders without props",
      props: {}
    },
    {
      desc: "renders with all props",
      props: {
        anchorPoint: anchorPoints.RIGHT_BOTTOM,
        heading: "Hello",
        indicatorTitle: "World",
        loading: true,
        notifications: [],
        onClickOutside: function handleClickOutside() {},
        onScroll: function handleScroll() {},
        open: true,
        unreadCount: 44
      }
    }
  ]);
});
