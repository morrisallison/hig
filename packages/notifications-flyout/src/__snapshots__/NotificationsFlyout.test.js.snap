// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notifications-flyout/NotificationsFlyout renders with all props 1`] = `
<div
  className="hig__flyout-v1 hig__flyout-v1--right-bottom"
>
  <div
    className="hig__flyout-v1__action"
  >
    <div
      className="hig__notifications-flyout__indicator"
    >
      <button
        className="hig__icon-button hig__icon-button--transparent"
        onBlur={undefined}
        onClick={[Function]}
        onFocus={undefined}
        onMouseEnter={undefined}
        onMouseLeave={undefined}
        title="World"
      >
        <span
          className="hig__icon-button__icon"
        >
          <div
            className="hig__icon hig__icon--24-size"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"4651 1106 24 24\\" role=\\"img\\" title=\\"notification-24\\"><path d=\\"M4669.931,1120.406l-0.112-0.05c-0.803-0.358-0.803-0.358-0.795-1.51l0.002-0.775c0.008-4.383-1.382-7.12-4.023-7.952V1109	c0-0.495-0.505-1-1-1h-2c-0.495,0-1,0.505-1,1v1.119c-2.653,0.833-4.051,3.571-4.047,7.952l0.002,0.778	c0.008,1.148,0.008,1.148-0.794,1.507l-0.111,0.05c-1.961,0.878-2.049,3.592-2.052,3.7c-0.009,0.248,0.087,0.49,0.253,0.656	c0.226,0.237,0.513,0.237,0.651,0.237H4660c0,1.654,1.346,3,3,3s3-1.346,3-3h5.078c0.137,0,0.42,0,0.647-0.233	c0.167-0.172,0.261-0.409,0.256-0.669C4671.962,1123.626,4671.77,1121.229,4669.931,1120.406z M4663,1127c-1.103,0-2-0.897-2-2h4	C4665,1126.103,4664.102,1127,4663,1127z M4655.009,1124c0.04-0.491,0.266-2.15,1.451-2.682l0.11-0.049	c1.171-0.522,1.399-0.77,1.387-2.425l-0.002-0.774c-0.002-2.838,0.631-6.37,3.66-7.084l0.387-0.091V1109h2v1.896l0.386,0.091	c3.016,0.71,3.643,4.244,3.638,7.083l-0.002,0.773c-0.011,1.657,0.218,1.904,1.389,2.427l0.11,0.049	c1.225,0.549,1.424,2.291,1.477,2.682H4655.009z\\"/></svg>",
              }
            }
          />
        </span>
      </button>
      <div
        className="hig__notifications-flyout__indicator__count"
      >
        44
      </div>
    </div>
  </div>
  <div
    className="hig__flyout-v1__container"
    style={
      Object {
        "left": -7,
        "top": -0,
      }
    }
  >
    <div
      aria-hidden="true"
      className="hig__flyout-v1__chevron"
      role="presentation"
    />
    <div
      aria-hidden="true"
      className="hig__flyout-v1__chevron-cover"
      role="presentation"
    />
    <div
      className="hig__flyout-v1__panel-container"
      style={
        Object {
          "maxHeight": undefined,
        }
      }
    >
      <div
        className="hig__flyout-v1__panel-container__inner"
      >
        <header
          className="hig__notifications-flyout__panel__title"
        >
          Hello
        </header>
        <section
          className="hig__notifications-flyout__panel__container"
          style={
            Object {
              "maxHeight": "688px",
            }
          }
        >
          <div
            onScroll={[Function]}
            role="list"
          />
        </section>
        <footer
          className="hig__notifications-flyout__panel__loading hig__notifications-flyout__panel__loading--open"
        >
          <div
            data-props="{
  \\"size\\": \\"s\\"
}"
          />
        </footer>
      </div>
    </div>
  </div>
</div>
`;

exports[`notifications-flyout/NotificationsFlyout renders without props 1`] = `
<div
  className="hig__flyout-v1 hig__flyout-v1--exited hig__flyout-v1--right-top"
>
  <div
    className="hig__flyout-v1__action"
  >
    <div
      className="hig__notifications-flyout__indicator"
    >
      <button
        className="hig__icon-button hig__icon-button--transparent"
        onBlur={undefined}
        onClick={[Function]}
        onFocus={undefined}
        onMouseEnter={undefined}
        onMouseLeave={undefined}
        title="View notifications"
      >
        <span
          className="hig__icon-button__icon"
        >
          <div
            className="hig__icon hig__icon--24-size"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"4651 1106 24 24\\" role=\\"img\\" title=\\"notification-24\\"><path d=\\"M4669.931,1120.406l-0.112-0.05c-0.803-0.358-0.803-0.358-0.795-1.51l0.002-0.775c0.008-4.383-1.382-7.12-4.023-7.952V1109	c0-0.495-0.505-1-1-1h-2c-0.495,0-1,0.505-1,1v1.119c-2.653,0.833-4.051,3.571-4.047,7.952l0.002,0.778	c0.008,1.148,0.008,1.148-0.794,1.507l-0.111,0.05c-1.961,0.878-2.049,3.592-2.052,3.7c-0.009,0.248,0.087,0.49,0.253,0.656	c0.226,0.237,0.513,0.237,0.651,0.237H4660c0,1.654,1.346,3,3,3s3-1.346,3-3h5.078c0.137,0,0.42,0,0.647-0.233	c0.167-0.172,0.261-0.409,0.256-0.669C4671.962,1123.626,4671.77,1121.229,4669.931,1120.406z M4663,1127c-1.103,0-2-0.897-2-2h4	C4665,1126.103,4664.102,1127,4663,1127z M4655.009,1124c0.04-0.491,0.266-2.15,1.451-2.682l0.11-0.049	c1.171-0.522,1.399-0.77,1.387-2.425l-0.002-0.774c-0.002-2.838,0.631-6.37,3.66-7.084l0.387-0.091V1109h2v1.896l0.386,0.091	c3.016,0.71,3.643,4.244,3.638,7.083l-0.002,0.773c-0.011,1.657,0.218,1.904,1.389,2.427l0.11,0.049	c1.225,0.549,1.424,2.291,1.477,2.682H4655.009z\\"/></svg>",
              }
            }
          />
        </span>
      </button>
      <div
        className="hig__notifications-flyout__indicator__count hig__notifications-flyout__indicator__count--hide"
      >
        0
      </div>
    </div>
  </div>
  <div
    className="hig__flyout-v1__container"
    style={
      Object {
        "left": -7,
        "top": 0,
      }
    }
  >
    <div
      aria-hidden="true"
      className="hig__flyout-v1__chevron"
      role="presentation"
    />
    <div
      aria-hidden="true"
      className="hig__flyout-v1__chevron-cover"
      role="presentation"
    />
    <div
      className="hig__flyout-v1__panel-container"
      style={
        Object {
          "maxHeight": undefined,
        }
      }
    >
      <div
        className="hig__flyout-v1__panel-container__inner"
      >
        <header
          className="hig__notifications-flyout__panel__title"
        >
          Notifications
        </header>
        <section
          className="hig__notifications-flyout__panel__container"
          style={
            Object {
              "maxHeight": "688px",
            }
          }
        >
          <div
            onScroll={undefined}
            role="list"
          />
        </section>
        <footer
          className="hig__notifications-flyout__panel__loading"
        >
          <div
            data-props="{
  \\"size\\": \\"s\\"
}"
          />
        </footer>
      </div>
    </div>
  </div>
</div>
`;
